<launch>

   <!-- Rosbag path -->
  <arg name="rosbag_path" default="/"  /> 
 
  <!-- Playback speed of rosbag -->
  <arg name="rosbag_rate" default="0.5"  />

    <!-- Parameters to run -->
  <arg name="rviz" default="true" />
  <arg name="lidar" default="MRS-6000" /> <!-- VLP-16  MRS-6000 -->
  <arg name="cloudTopic" default="/sick_mrs6xxx/cloud" /> <!-- /velodyne/velodyne_points  /sick_mrs6xxx/cloud  /sick_mrs6xxx/cloud_single_echo -->
  <arg name="scanPeriod" default="0.1" />  
  
  <param name="use_sim_time" value="true"/>

    <!-- MapManager parameters - MapManager must be disabled/enabled in code -->
  <arg name="features_minDstPtToPl" default="0.3" /> 
  <arg name="features_minDstPtToPt" default="1.1" /> 
  <arg name="matching_minDstPtToPl" default="0.5" /> 
  <arg name="matching_minDstPtToPt" default="0.6" />
  <arg name="deleting1_PtNum" default="15" />   
  <arg name="deleting2_PtNum" default="25" /> 
  <arg name="deleting2_iterations" default="2" />  
  
   <!-- Starts rosbag -->
   <node pkg="rosbag" type="play" name="rosbag_play"
          args="$(arg rosbag_path) --clock --delay 5 -r $(arg rosbag_rate) -s 0" output="screen" />


  <!-- Starts multiScanRegistration -->
  <node pkg="loam_velodyne" type="multiScanRegistration" name="multiScanRegistration" output="screen" >
    <param name="lidar" value="$(arg lidar)" /> <!-- options: VLP-16  HDL-32  HDL-64E MRS-6000 -->
    <param name="scanPeriod" value="$(arg scanPeriod)" />
    <remap from="/multi_scan_points" to="$(arg cloudTopic)" />
  </node>
 
  <!-- Starts laserOdometry -->
  <node pkg="loam_velodyne" type="laserOdometry" name="laserOdometry" output="screen" respawn="true">
    <param name="scanPeriod" value="$(arg scanPeriod)" />

  </node>
  
  <!-- Starts laserMapping -->
  <node pkg="loam_velodyne" type="laserMapping" name="laserMapping" output="screen">
    <param name="scanPeriod" value="$(arg scanPeriod)" />

     <!-- MapManager parameters -->
    <param name="features_minDstPtToPl" value="$(arg features_minDstPtToPl)" />
    <param name="features_minDstPtToPt" value="$(arg features_minDstPtToPt)" />
    <param name="matching_minDstPtToPl" value="$(arg matching_minDstPtToPl)" />
    <param name="matching_minDstPtToPt" value="$(arg matching_minDstPtToPt)" />
    <param name="deleting1_PtNum" value="$(arg deleting1_PtNum)" />
    <param name="deleting2_PtNum" value="$(arg deleting2_PtNum)" />
    <param name="deleting2_iterations" value="$(arg deleting2_iterations)" />

  </node>
  
  <!-- Starts tranfsormMaintenance -->
  <node pkg="loam_velodyne" type="transformMaintenance" name="transformMaintenance" output="screen">
  </node>

  <!-- Starts Rviz -->
  <group if="$(arg rviz)">
    <node launch-prefix="nice" pkg="rviz" type="rviz" name="rviz" args="-d $(find loam_velodyne)/rviz_cfg/feature_based_loam.rviz" />
  </group>


</launch>
